### ISSUE TITLE
HELP! 'NoneType' object has no attribute 'next_functions'

### ISSUE THREAD

**Issue by galvinator** [ThilinaRajapakse/simpletransformers#582]
@ThilinaRajapakse Thank you for creating this library. It's really useful. However I encountered the following issue while implementing your library. Can you help me out in it? Apologies for bad editing. I'm new to github. Also I run this code on Google Colab with GPU enabled. For context, the data set I am using has approximately 3000+ rows, but there are only 66 distinct labels. The format of `train_data` and `eval_data` are `Dataframe`.

<pre><code>
from simpletransformers.classification import ClassificationModel, ClassificationArgs
import pandas as pd
import logging

logging.basicConfig(level=logging.INFO)
transformers_logger = logging.getLogger("transformers")
transformers_logger.setLevel(logging.WARNING)

# Optional model configuration
model_args = ClassificationArgs(num_train_epochs=1)

# Create a ClassificationModel
model = ClassificationModel(
    "electra",
    "google/electra-base-discriminator",
    num_labels=65,
    #use_cuda = False,
    args=model_args
) 

# Train the model
model.train_model(train_data)

# Evaluate the model
result, model_outputs, wrong_predictions = model.eval_model(eval_data)Jupyter output')

# Make predictions with the model
predictions, raw_outputs = model.predict(["computer programming"])
</code></pre>

**Below is the error message:**
<pre><code>
WARNING:transformers.modeling_utils:Some weights of the model checkpoint at google/electra-base-discriminator were not used when initializing ElectraForSequenceClassification: ['discriminator_predictions.dense.weight', 'discriminator_predictions.dense.bias', 'discriminator_predictions.dense_prediction.weight', 'discriminator_predictions.dense_prediction.bias', 'electra.embeddings_project.weight', 'electra.embeddings_project.bias']
- This IS expected if you are initializing ElectraForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPretraining model).
- This IS NOT expected if you are initializing ElectraForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
WARNING:transformers.modeling_utils:Some weights of ElectraForSequenceClassification were not initialized from the model checkpoint at google/electra-base-discriminator and are newly initialized: ['pooler.dense.weight', 'pooler.dense.bias', 'classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
INFO:simpletransformers.classification.classification_model: Converting to features started. Cache is not used.
100%
2530/2530 [00:02<00:00, 907.60it/s]

Epoch 1 of 1: 0%
0/1 [00:05<?, ?it/s]
Running Epoch 0 of 1: 0%
0/317 [00:05<?, ?it/s]

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-56-dfc339c0df49> in <module>()
     39 
     40 # Train the model
---> 41 model.train_model(train_data)#, output_dir = r'D:\Academia\UNI\RIPS\ELECTRA Jupyter output')
     42 
     43 # Evaluate the model

11 frames
/usr/local/lib/python3.6/dist-packages/simpletransformers/classification/classification_model.py in train_model(self, train_df, multi_label, output_dir, show_running_loss, args, eval_df, verbose, **kwargs)
    294             eval_df=eval_df,
    295             verbose=verbose,
--> 296             **kwargs,
    297         )
    298 

/usr/local/lib/python3.6/dist-packages/simpletransformers/classification/classification_model.py in train(self, train_dataloader, output_dir, multi_label, show_running_loss, eval_df, verbose, **kwargs)
    459 
    460                 inputs = self._get_inputs_dict(batch)
--> 461                 outputs = model(**inputs)
    462                 # model outputs are always tuple in pytorch-transformers (see doc)
    463                 loss = outputs[0]

/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    548             result = self._slow_forward(*input, **kwargs)
    549         else:
--> 550             result = self.forward(*input, **kwargs)
    551         for hook in self._forward_hooks.values():
    552             hook_result = hook(self, input, result)

/usr/local/lib/python3.6/dist-packages/simpletransformers/custom_models/models.py in forward(self, input_ids, attention_mask, token_type_ids, position_ids, head_mask, inputs_embeds, labels)
    533         outputs = self.electra(input_ids, attention_mask, token_type_ids, position_ids, head_mask, inputs_embeds)
    534         sequence_output = outputs[0]
--> 535         pooled_output = self.pooler(sequence_output)
    536         pooled_output = self.dropout(pooled_output)
    537         logits = self.classifier(pooled_output)

/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    548             result = self._slow_forward(*input, **kwargs)
    549         else:
--> 550             result = self.forward(*input, **kwargs)
    551         for hook in self._forward_hooks.values():
    552             hook_result = hook(self, input, result)

/usr/local/lib/python3.6/dist-packages/simpletransformers/custom_models/models.py in forward(self, hidden_states)
    413         # to the first token.
    414         first_token_tensor = hidden_states[:, 0]
--> 415         pooled_output = self.dense(first_token_tensor)
    416         pooled_output = self.activation(pooled_output)
    417         return pooled_output

/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    548             result = self._slow_forward(*input, **kwargs)
    549         else:
--> 550             result = self.forward(*input, **kwargs)
    551         for hook in self._forward_hooks.values():
    552             hook_result = hook(self, input, result)

/usr/local/lib/python3.6/dist-packages/torch/nn/modules/linear.py in forward(self, input)
     85 
     86     def forward(self, input):
---> 87         return F.linear(input, self.weight, self.bias)
     88 
     89     def extra_repr(self):

/usr/local/lib/python3.6/dist-packages/apex/amp/wrap.py in wrapper(*args, **kwargs)
     26                                      args,
     27                                      kwargs)
---> 28         return orig_fn(*new_args, **kwargs)
     29     return wrapper
     30 

/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py in linear(input, weight, bias)
   1608     if input.dim() == 2 and bias is not None:
   1609         # fused op is marginally faster
-> 1610         ret = torch.addmm(bias, input, weight.t())
   1611     else:
   1612         output = input.matmul(weight.t())

/usr/local/lib/python3.6/dist-packages/apex/amp/wrap.py in wrapper(*args, **kwargs)
     19             for i in range(len(args)):
     20                 if utils.should_cache(args[i]):
---> 21                     args[i] = utils.cached_cast(cast_fn, args[i], handle.cache)
     22             for k in kwargs:
     23                 if utils.should_cache(kwargs[k]):

/usr/local/lib/python3.6/dist-packages/apex/amp/utils.py in cached_cast(cast_fn, x, cache)
     95         if x.requires_grad and cached_x.requires_grad:
     96             # Make sure x is actually cached_x's autograd parent.
---> 97             if cached_x.grad_fn.next_functions[1][0].variable is not x:
     98                 raise RuntimeError("x and cache[x] both require grad, but x is not "
     99                                    "cache[x]'s parent.  This is likely an error.")

AttributeError: 'NoneType' object has no attribute 'next_functions'
</code></pre>

**Comment by galvinator** [comment]
On the other hand, if I run the same script, but with GPU turned off (only CPU) and use_cuda = False, I encounter yet another set of error, 

<pre><code>
INFO:filelock:Lock 140422570599480 acquired on /root/.cache/torch/transformers/9236d197566a7f1be2b2151f5afcc5a8e17f31e1e23c52f3cdf2340019986e78.88ba6e8e7d5a7936e86d6f2551fe19c236dc57c24da163907cd0544e9933f6ee.lock
Downloading: 100%
467/467 [00:09<00:00, 50.8B/s]
INFO:filelock:Lock 140422570599480 released on /root/.cache/torch/transformers/9236d197566a7f1be2b2151f5afcc5a8e17f31e1e23c52f3cdf2340019986e78.88ba6e8e7d5a7936e86d6f2551fe19c236dc57c24da163907cd0544e9933f6ee.lock

INFO:filelock:Lock 140422570601104 acquired on /root/.cache/torch/transformers/1e208ba81f889d92ba15c70c4c400416d5ac7a2d017a6be053d69a5c41b9c6af.b7514d01ce5acfe02313470cce3175018852a5e8cbcb8784268ab87dc21daf4c.lock
Downloading: 100%
440M/440M [00:08<00:00, 49.7MB/s]
INFO:filelock:Lock 140422570601104 released on /root/.cache/torch/transformers/1e208ba81f889d92ba15c70c4c400416d5ac7a2d017a6be053d69a5c41b9c6af.b7514d01ce5acfe02313470cce3175018852a5e8cbcb8784268ab87dc21daf4c.lock

WARNING:transformers.modeling_utils:Some weights of the model checkpoint at google/electra-base-discriminator were not used when initializing ElectraForSequenceClassification: ['discriminator_predictions.dense.weight', 'discriminator_predictions.dense.bias', 'discriminator_predictions.dense_prediction.weight', 'discriminator_predictions.dense_prediction.bias', 'electra.embeddings_project.weight', 'electra.embeddings_project.bias']
- This IS expected if you are initializing ElectraForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPretraining model).
- This IS NOT expected if you are initializing ElectraForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
WARNING:transformers.modeling_utils:Some weights of ElectraForSequenceClassification were not initialized from the model checkpoint at google/electra-base-discriminator and are newly initialized: ['pooler.dense.weight', 'pooler.dense.bias', 'classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
INFO:filelock:Lock 140422557113704 acquired on /root/.cache/torch/transformers/ff085885d4c95651587af553adadd34a26de8a663f2cef709635b48b3bed2bbd.542ce4285a40d23a559526243235df47c5f75c197f04f37d1a0c124c32c9a084.lock
Downloading: 100%
232k/232k [00:00<00:00, 1.49MB/s]
INFO:filelock:Lock 140422557113704 released on /root/.cache/torch/transformers/ff085885d4c95651587af553adadd34a26de8a663f2cef709635b48b3bed2bbd.542ce4285a40d23a559526243235df47c5f75c197f04f37d1a0c124c32c9a084.lock
INFO:simpletransformers.classification.classification_model: Converting to features started. Cache is not used.

100%
2530/2530 [00:02<00:00, 1115.49it/s]

Epoch 1 of 1: 0%
0/1 [00:40<?, ?it/s]
Epochs 0/1. Running Loss: 4.2326: 1%
3/317 [00:40<1:10:28, 13.47s/it]


---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-10-dfc339c0df49> in <module>()
     39 
     40 # Train the model
---> 41 model.train_model(train_data)#, output_dir = r'D:\Academia\UNI\RIPS\ELECTRA Jupyter output')
     42 
     43 # Evaluate the model

7 frames
/usr/local/lib/python3.6/dist-packages/simpletransformers/classification/classification_model.py in train_model(self, train_df, multi_label, output_dir, show_running_loss, args, eval_df, verbose, **kwargs)
    294             eval_df=eval_df,
    295             verbose=verbose,
--> 296             **kwargs,
    297         )
    298 

/usr/local/lib/python3.6/dist-packages/simpletransformers/classification/classification_model.py in train(self, train_dataloader, output_dir, multi_label, show_running_loss, eval_df, verbose, **kwargs)
    459 
    460                 inputs = self._get_inputs_dict(batch)
--> 461                 outputs = model(**inputs)
    462                 # model outputs are always tuple in pytorch-transformers (see doc)
    463                 loss = outputs[0]

/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    548             result = self._slow_forward(*input, **kwargs)
    549         else:
--> 550             result = self.forward(*input, **kwargs)
    551         for hook in self._forward_hooks.values():
    552             hook_result = hook(self, input, result)

/usr/local/lib/python3.6/dist-packages/simpletransformers/custom_models/models.py in forward(self, input_ids, attention_mask, token_type_ids, position_ids, head_mask, inputs_embeds, labels)
    545             else:
    546                 loss_fct = CrossEntropyLoss(weight=self.weight)
--> 547                 loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
    548             outputs = (loss,) + outputs
    549 

/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    548             result = self._slow_forward(*input, **kwargs)
    549         else:
--> 550             result = self.forward(*input, **kwargs)
    551         for hook in self._forward_hooks.values():
    552             hook_result = hook(self, input, result)

/usr/local/lib/python3.6/dist-packages/torch/nn/modules/loss.py in forward(self, input, target)
    930     def forward(self, input, target):
    931         return F.cross_entropy(input, target, weight=self.weight,
--> 932                                ignore_index=self.ignore_index, reduction=self.reduction)
    933 
    934 

/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py in cross_entropy(input, target, weight, size_average, ignore_index, reduce, reduction)
   2315     if size_average is not None or reduce is not None:
   2316         reduction = _Reduction.legacy_get_string(size_average, reduce)
-> 2317     return nll_loss(log_softmax(input, 1), target, weight, None, ignore_index, None, reduction)
   2318 
   2319 

/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py in nll_loss(input, target, weight, size_average, ignore_index, reduce, reduction)
   2113                          .format(input.size(0), target.size(0)))
   2114     if dim == 2:
-> 2115         ret = torch._C._nn.nll_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index)
   2116     elif dim == 4:
   2117         ret = torch._C._nn.nll_loss2d(input, target, weight, _Reduction.get_enum(reduction), ignore_index)

IndexError: Target 65 is out of bounds.
<code><pre>

**Comment by ThilinaRajapakse** [comment]
It looks like Apex wasn't installed correctly. Can you try running it without `fp16`?

```python
# Optional model configuration
model_args = ClassificationArgs(num_train_epochs=1, fp16=False)
```

This will likely result in the same error you see when running on the CPU. That error seems to be caused by `num_labels` being too low or because the labels in your data start from 1 and go to 65 (included). Setting `num_labels=65` means that your labels can be anything from 0 (included) up to but _not including_ 65.

**Comment by galvinator** [comment]
@ThilinaRajapakse I ran without `fp16`, changing the value of `num_labels` from 65 to 66 and running on CPU - seems like no issues although the estimated training time is about 1.5 hours. I will try it again later in the day (it's 3am here)

I believe installed Apex properly. I ran (as instructed on [https://github.com/NVIDIA/apex#quick-start])
<pre><code>
!git clone https://github.com/NVIDIA/apex
!cd apex
!pip install -v --no-cache-dir --global-option="--cpp_ext" --global-option="--cuda_ext" /content/apex
<code><pre>
on colab and obtained these messages at the end
<pre><code>
Successfully installed apex-0.1
Cleaning up...
Removed build tracker '/tmp/pip-req-tracker-va3m4uvf'
<pre><code>. 

I will try running the code on Colab with GPU enabled,` fp16 = True` and `use_cuda = True` later too, but should I be concerned of anything else? I really hope to run my code using GPU than CPU.

**Comment by galvinator** [comment]
@ThilinaRajapakse Hi, it works after changing the number of labels. Once again, thank you so much for your assistance and guidance. Superb library you got there!

**Comment by ThilinaRajapakse** [comment]
Awesome! And thank you!

**Comment by galvinator** [comment]
@ThilinaRajapakse 

Little help again, I am trying to use wandb's hyperparameterization tuning `sweep` method. 
This is my code,
`import sklearn 
from simpletransformers.classification import ClassificationModel, ClassificationArgs
import pandas as pd
import logging

sweep_config = {
    "method": "bayes",  # grid, random
    "metric": {"name": "train_loss", "goal": "minimize"},
    "parameters": {
        "num_train_epochs": {"values": [2, 3, 5]},
        "learning_rate": {"min": 5e-5, "max": 1e-3},
        "train_batch_size":{"values":[8,16,32,64,128]}
    },
}

sweep_id = wandb.sweep(sweep_config, project="Simple Sweep")

logging.basicConfig(level=logging.INFO)
transformers_logger = logging.getLogger("transformers")
transformers_logger.setLevel(logging.WARNING)

# Optional model configuration
model_args = ClassificationArgs(sliding_window = True)#, fp16 = False)
model_args.num_train_epochs = 5
model_args.learning_rate = 1e-4
model_args.train_batch_size = 64
model_args.overwrite_output_dir = True
model_args.reprocess_input_data = True
model_args.no_save = True
use_multiprocessing = False
###########################################
model_args.use_early_stopping = True
model_args.early_stopping_delta = 0.01
model_args.early_stopping_metric = "mcc"
model_args.early_stopping_metric_minimize = False
model_args.early_stopping_patience = 5
model_args.evaluate_during_training_steps = 1000
model_args.evaluate_during_training = True
###########################################

def train():
    # Initialize a new wandb run
    wandb.init()

    # Create a TransformerModel
    model = ClassificationModel(
        "electra",
        "google/electra-base-discriminator",
        args=model_args,
        sweep_config=wandb.config,
    )

    # Train the model
    model.train_model(train_data, eval_df=eval_data)

    # Evaluate the model
    model.eval_model(eval_data)

    # Sync wandb
    wandb.join()


wandb.agent(sweep_id, train)`

However I obtained the following error, 
`Create sweep with ID: xny1rj35
Sweep URL: https://app.wandb.ai/galvinator/Simple%20Sweep/sweeps/xny1rj35
INFO:wandb.wandb_agent:Running runs: []
INFO:wandb.wandb_agent:Agent received command: run
INFO:wandb.wandb_agent:Agent starting run with config:
	learning_rate: 0.0006286912410153943
	num_train_epochs: 3
	train_batch_size: 128
wandb: Agent Starting Run: 3u0uw3l3 with config:
	learning_rate: 0.0006286912410153943
	num_train_epochs: 3
	train_batch_size: 128
wandb: Agent Started Run: 3u0uw3l3
Logging results to Weights & Biases (Documentation).
Project page: https://app.wandb.ai/galvinator/Simple%20Sweep
Sweep page: https://app.wandb.ai/galvinator/Simple%20Sweep/sweeps/xny1rj35
Run page: https://app.wandb.ai/galvinator/Simple%20Sweep/runs/3u0uw3l3
INFO:wandb.run_manager:system metrics and metadata threads started
INFO:wandb.run_manager:checking resume status, waiting at most 10 seconds
INFO:wandb.run_manager:resuming run from id: UnVuOnYxOjN1MHV3M2wzOlNpbXBsZSBTd2VlcDpnYWx2aW5hdG9y
INFO:wandb.run_manager:upserting run before process can begin, waiting at most 10 seconds
INFO:wandb.run_manager:saving pip packages
INFO:wandb.run_manager:initializing streaming files api
INFO:wandb.run_manager:unblocking file change observer, beginning sync with W&B servers
INFO:wandb.run_manager:file/dir modified: /content/wandb/run-20200721_141835-3u0uw3l3/config.yaml
INFO:wandb.run_manager:file/dir created: /content/wandb/run-20200721_141835-3u0uw3l3/wandb-metadata.json
INFO:wandb.run_manager:file/dir created: /content/wandb/run-20200721_141835-3u0uw3l3/requirements.txt
INFO:wandb.run_manager:file/dir created: /content/wandb/run-20200721_141835-3u0uw3l3/wandb-events.jsonl
INFO:wandb.run_manager:file/dir created: /content/wandb/run-20200721_141835-3u0uw3l3/wandb-summary.json
INFO:wandb.run_manager:file/dir created: /content/wandb/run-20200721_141835-3u0uw3l3/wandb-history.jsonl
WARNING:transformers.modeling_utils:Some weights of the model checkpoint at google/electra-base-discriminator were not used when initializing ElectraForSequenceClassification: ['discriminator_predictions.dense.weight', 'discriminator_predictions.dense.bias', 'discriminator_predictions.dense_prediction.weight', 'discriminator_predictions.dense_prediction.bias', 'electra.embeddings_project.weight', 'electra.embeddings_project.bias']
- This IS expected if you are initializing ElectraForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPretraining model).
- This IS NOT expected if you are initializing ElectraForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
WARNING:transformers.modeling_utils:Some weights of ElectraForSequenceClassification were not initialized from the model checkpoint at google/electra-base-discriminator and are newly initialized: ['pooler.dense.weight', 'pooler.dense.bias', 'classifier.weight', 'classifier.bias']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Process Process-2:
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.6/dist-packages/wandb/wandb_agent.py", line 64, in _start
    function()
  File "<ipython-input-18-bfff4d68276b>", line 54, in train
    model.train_model(train_data, eval_df=eval_data)
  File "/usr/local/lib/python3.6/dist-packages/simpletransformers/classification/classification_model.py", line 252, in train_model
    self._move_model_to_device()
  File "/usr/local/lib/python3.6/dist-packages/simpletransformers/classification/classification_model.py", line 1169, in _move_model_to_device
    self.model.to(self.device)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 443, in to
    return self._apply(convert)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 203, in _apply
    module._apply(fn)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 203, in _apply
    module._apply(fn)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 203, in _apply
    module._apply(fn)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 225, in _apply
    param_applied = fn(param)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 441, in convert
    return t.to(device, dtype if t.is_floating_point() else None, non_blocking)
  File "/usr/local/lib/python3.6/dist-packages/torch/cuda/__init__.py", line 148, in _lazy_init
    "Cannot re-initialize CUDA in forked subprocess. " + msg)
RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method
INFO:wandb.wandb_agent:Running runs: ['3u0uw3l3']`

I'm not very familiar with deep learning models and their implementation. Might you assist me in this problem? Thanks

**Comment by galvinator** [comment]
Hi @ThilinaRajapakse I tried using `wandb` method sweep` for hyperparameter optimization. The following is my code

```
from torch.multiprocessing import Pool, Process, set_start_method
try:
     torch.multiprocessing.set_start_method('spawn', force=True)
except RuntimeError:
    pass

import sklearn 
from simpletransformers.classification import ClassificationModel, ClassificationArgs
import pandas as pd
import logging

sweep_config = {
    "method": "bayes",  # grid, random
    "metric": {"name": "accuracy", "goal": "maximize"},
    "parameters": {
        "num_train_epochs": {"min":1,"max":40},
        "learning_rate": {"min": 5e-5, "max": 1e-3},
        "train_batch_size":{"values":[8,16,32,64,128]}
    },
    #W&B Sweeps can also speed up the hyperparameter optimization by terminating any poorly performing runs
    "early_terminate": {"type": "hyperband", "min_iter": 6,}
}

sweep_id = wandb.sweep(sweep_config, project="Simple Sweep")

logging.basicConfig(level=logging.INFO)
transformers_logger = logging.getLogger("transformers")
transformers_logger.setLevel(logging.WARNING)

# Optional model configuration
model_args = ClassificationArgs(sliding_window = True)#, fp16 = False)
# model_args.num_train_epochs = 5
# model_args.learning_rate = 1e-4
# model_args.train_batch_size = 64
model_args.overwrite_output_dir = True
model_args.reprocess_input_data = True
model_args.no_save = True
use_multiprocessing = False
###########################################
model_args.use_early_stopping = True
model_args.early_stopping_delta = 0.01
model_args.early_stopping_metric = "mcc"
model_args.early_stopping_metric_minimize = False
model_args.early_stopping_patience = 5
model_args.evaluate_during_training_steps = 1000
model_args.evaluate_during_training = True
model_args.wandb_project = "Simple Sweep"
###########################################

def train():
    # Initialize a new wandb run
    wandb.init()

    # Create a TransformerModel
    model = ClassificationModel(
        "electra",
        "google/electra-base-discriminator",
        args=model_args,
        sweep_config=wandb.config,
    )

    # Train the model
    model.train_model(train_data, eval_df=eval_data)

    # Evaluate the model
    model.eval_model(eval_data)

    # Sync wandb
    wandb.join()


wandb.agent(sweep_id, train)
```

The following message is what I observe 
```
Create sweep with ID: 7ec5a8d3
Sweep URL: https://app.wandb.ai/galvinator/Simple%20Sweep/sweeps/7ec5a8d3
INFO:wandb.wandb_agent:Running runs: []
INFO:wandb.wandb_agent:Agent received command: run
INFO:wandb.wandb_agent:Agent starting run with config:
	learning_rate: 0.0002931921479743678
	num_train_epochs: 29
	train_batch_size: 128
wandb: Agent Starting Run: i7dnig85 with config:
	learning_rate: 0.0002931921479743678
	num_train_epochs: 29
	train_batch_size: 128
INFO:wandb.wandb_agent:Running runs: ['i7dnig85']
```

Is that supposed to be expected? It's a bit unusual and the process is still running.

**Comment by ThilinaRajapakse** [comment]
If there are models being trained, then yes.

**Comment by galvinator** [comment]
It doesn't seem to be iterating through and is stuck on just one combination of model_args for 1 hour+, 
that is the combination of 
`
learning_rate: 0.0002931921479743678
num_train_epochs: 29
train_batch_size: 128
`
I don't think that is supposed to happen, right? 
Shouldn't `sweep` be trying other permutations of parameters?

**Comment by ThilinaRajapakse** [comment]
I can't spot anything wrong in the code.

Once the first run completes, the second run should definitely start with different hyperparameter values. What do you get on the Sweep dashboard on W&B?

**Comment by galvinator** [comment]
That's what I thought so too. For the Sweep dashboard on W&B, I only see one graph with a curve, something like this:
![image](https://user-images.githubusercontent.com/66785682/88216964-70ea9480-cc90-11ea-8814-6ea3e9606f3b.png)

for one hour plus.. Perhaps, I might have been too impatient as I tried running a simple training with num_of_epochs = 29 and learning rate of 1e-4 and the training time was close to 2 hours.

**Comment by ThilinaRajapakse** [comment]
That means that the first run most likely didn't complete. You can try running with a small subset of your actual data to verify that the Sweep is actually working. In fact, it's always a good idea to run the training with a small subset first to check that the code runs as expected (and to completion).

**Comment by galvinator** [comment]
Yes I plan to try that out. Thanks for your guidance!
